### ISSUE TITLE
[ENHANCEMENT]: Provide helper function cuco::experimental::make_static_set et al

### ISSUE THREAD

**Issue by esoha-nvidia** [NVIDIA/cuCollections#370]
### Is your feature request related to a problem? Please describe.

I would like a function called `make_static_set` that will create a static set that will infer the template arguments from the function call.

### Describe the solution you'd like

Instead of:

```c++
auto set = cuco::experimental::static_set<MyKeyType, MyExtent, MyThreadScope, MyKeyEqual>(capacity, cuco::empty_key(-1), MyKeyEqual(123));
```

I could write:

```c++
auto set = cuco::experimental::make_static_set(capacity, cuco::empty_key(-1), MyKeyEqual(123));
```

Notice how it's much smaller and I'm not repeating myself.  Notice the addition of the work "make".  This is consistent with the stdlib's usage of the word make in, for example, `std::make_tuple`

### Describe alternatives you've considered

I can do without but its wordy.  Doing this doesn't add much work for cuco.  Cuco can modify some of its tests to use this form of the instantiation so that it is tested well in CI.

### Additional context

Inference only works on functions, not classes: https://stackoverflow.com/questions/797594/when-a-compiler-can-infer-a-template-parameter

Here's how `make_tuple` works: https://en.cppreference.com/w/cpp/utility/tuple/make_tuple

**Comment by jrhemstad** [comment]
> Inference only works on functions, not classes: https://stackoverflow.com/questions/797594/when-a-compiler-can-infer-a-template-parameter

This is no longer true as of C++17 with "Class Template Argument Deduction" (CTAD). CTAD has all but made `make_*` functions obsolete.

**Comment by sleeepyjack** [comment]
Yes, the error you are seeing is due to us struggling with getting CTAD right 🙃 
I wonder if this is something we can fix with a deduction guide?

**Comment by PointKernel** [comment]
> Yes, the error you are seeing is due to us struggling with getting CTAD right 🙃 I wonder if this is something we can fix with a deduction guide?

There are a few things in my mind to improve the situation. Wait for my PR(s)

**Comment by esoha-nvidia** [comment]
I can test it very easily if you need.  Let me know.

**Comment by PointKernel** [comment]
https://godbolt.org/z/4Y4WYfxcK

@esoha-nvidia actually, your example would deduce just fine even with the current implementation. Any other examples you would like to make them work?

**Comment by PointKernel** [comment]
Updates: now with #346 being merged. Plain integers can be deduced as well.

https://godbolt.org/z/Wsbr17fh5
